### PR TITLE
Reland add `write_record_metadata` to PyTorchFileWriter

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Reland of https://github.com/pytorch/pytorch/pull/125184 with compiler warning fixed by extending `m_pWrite` rather than adding `m_pSeek` to miniz API

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126087

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D57287327](https://our.internmc.facebook.com/intern/diff/D57287327)

### CODE REVIEW

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57287327).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57287327).

File: torch/csrc/jit/python/init.cpp [hunk 1]
```diff
@@ -1388,15 +1388,37 @@ void initJITBindings(PyObject* module) {
...
-          py::gil_scoped_acquire acquire;
-          auto memory_view = py::memoryview::from_memory(
-              reinterpret_cast<const char*>(data), size);
-          buffer.attr("write")(std::move(memory_view));
+          if (!data) {
+            // See [Note: write_record_metadata]
+            buffer.attr("seek")(
+                size, py::module::import("os").attr("SEEK_CUR"));
+          } else {
+            py::gil_scoped_acquire acquire;
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:1396]: You need the gil for any python api! So the code above needs it as well!

**Review by albanD — COMMENTED** [review]
GIL needs fixing, but SGTM otherwise!

File: torch/csrc/jit/python/init.cpp (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/csrc/jit/python/init.cpp]: Oops, very much my bad! Thanks for catching this, fixed

**Review by albanD — APPROVED** [review]
Ok!

**Pull request closed without merging**